File: sdk/rust/src/crypto.rs
use std::path::Path;

use base64::{engine::general_purpose::STANDARD, Engine};
use x25519_dalek::{PublicKey as XPublicKey, StaticSecret};

use crate::Error;

#[derive(Clone, Debug)]
pub struct PublicKey(pub(crate) XPublicKey);

#[derive(Clone)]
pub struct PrivateKey(StaticSecret);

impl PrivateKey {
    pub fn generate() -> Self {
        Self(StaticSecret::random())
    }

    pub fn public_key(&self) -> PublicKey {
        PublicKey(XPublicKey::from(&self.0))
    }

    pub fn to_bytes(&self) -> [u8; 32] {
        self.0.to_bytes()
    }

    pub fn to_base64(&self) -> String {
        STANDARD.encode(self.0.to_bytes())
    }

    pub fn from_base64(s: &str) -> Result<Self, Error> {
        let raw = STANDARD.decode(s).map_err(|e| Error::Crypto(e.to_string()))?;
        let arr: [u8; 32] = raw
            .try_into()
            .map_err(|_| Error::Crypto("private key must be 32 bytes".into()))?;
        Ok(Self(StaticSecret::from(arr)))
    }

    pub fn save_to_file(&self, path: &Path) -> Result<(), Error> {
        if let Some(parent) = path.parent() {
            std::fs::create_dir_all(parent)?;
        }
        std::fs::write(path, format!("{}\n", self.to_base64()))?;
        #[cfg(unix)]
        {
            use std::os::unix::fs::PermissionsExt;
            std::fs::set_permissions(path, std::fs::Permissions::from_mode(0o600))?;
        }
        Ok(())
    }

    pub fn load_from_file(path: &Path) -> Result<Self, Error> {
        let s = std::fs::read_to_string(path)?;
        Self::from_base64(s.trim())
    }

    pub fn load_or_generate(path: &Path) -> Result<Self, Error> {
        if path.exists() {
            return Self::load_from_file(path);
        }
        let key = Self::generate();
        key.save_to_file(path)?;
        Ok(key)
    }
}

impl PublicKey {
    pub fn to_bytes(&self) -> [u8; 32] {
        self.0.to_bytes()
    }

    pub fn to_base64(&self) -> String {
        STANDARD.encode(self.0.as_bytes())
    }

    pub fn from_base64(s: &str) -> Result<Self, Error> {
        let raw = STANDARD.decode(s).map_err(|e| Error::Crypto(e.to_string()))?;
        Self::from_bytes(&raw)
    }

    pub fn from_bytes(raw: &[u8]) -> Result<Self, Error> {
        let arr: [u8; 32] = raw
            .try_into()
            .map_err(|_| Error::Crypto("public key must be 32 bytes".into()))?;
        Ok(Self(XPublicKey::from(arr)))
    }
}

impl PartialEq for PublicKey {
    fn eq(&self, other: &Self) -> bool {
        self.0.as_bytes() == other.0.as_bytes()
    }
}

pub fn generate_key() -> PrivateKey {
    PrivateKey::generate()
}

pub fn load_key_from_file(path: &Path) -> Result<PrivateKey, Error> {
    PrivateKey::load_from_file(path)
}

pub fn load_or_generate_key(path: &Path) -> Result<PrivateKey, Error> {
    PrivateKey::load_or_generate(path)
}

File: sdk/cpp/src/crypto.cpp
#include "relayly/crypto.hpp"

#include <sodium.h>

#include <filesystem>
#include <fstream>
#include <sstream>

#include "relayly/errors.hpp"

#ifndef _WIN32
#include <sys/stat.h>
#endif

namespace relayly {

namespace {

namespace fs = std::filesystem;

std::string Base64Encode(const std::array<std::uint8_t, 32>& bytes) {
  // sodium_base64_ENCODED_LEN includes the null terminator.
  std::string out(sodium_base64_ENCODED_LEN(bytes.size(), sodium_base64_VARIANT_ORIGINAL), '\0');
  sodium_bin2base64(out.data(), out.size(), bytes.data(), bytes.size(), sodium_base64_VARIANT_ORIGINAL);
  out.resize(out.size() - 1);  // drop the trailing '\0' baked into the buffer size
  return out;
}

std::array<std::uint8_t, 32> Base64Decode(const std::string& b64, ErrorCode error_context) {
  std::array<std::uint8_t, 32> out{};
  std::size_t decoded_len = 0;
  if (sodium_base642bin(out.data(), out.size(), b64.data(), b64.size(), nullptr, &decoded_len, nullptr,
                         sodium_base64_VARIANT_ORIGINAL) != 0 ||
      decoded_len != out.size()) {
    throw Error(error_context, "key must be 32 bytes of base64");
  }
  return out;
}

std::string Trim(const std::string& s) {
  auto begin = s.find_first_not_of(" \t\r\n");
  if (begin == std::string::npos) return "";
  auto end = s.find_last_not_of(" \t\r\n");
  return s.substr(begin, end - begin + 1);
}

}  // namespace

PublicKey PublicKey::FromBase64(const std::string& b64) { return PublicKey(Base64Decode(b64, ErrorCode::kCrypto)); }

std::string PublicKey::ToBase64() const { return Base64Encode(bytes_); }

PrivateKey PrivateKey::Generate() {
  std::array<std::uint8_t, 32> priv{};
  std::array<std::uint8_t, 32> pub{};
  crypto_box_keypair(pub.data(), priv.data());
  return PrivateKey(priv);
}

PrivateKey PrivateKey::FromBase64(const std::string& b64) { return PrivateKey(Base64Decode(b64, ErrorCode::kCrypto)); }

PrivateKey PrivateKey::LoadFromFile(const std::string& path) {
  std::ifstream in(path);
  if (!in) {
    throw Error(ErrorCode::kIo, "failed to open key file: " + path);
  }
  std::ostringstream contents;
  contents << in.rdbuf();
  return FromBase64(Trim(contents.str()));
}

PrivateKey PrivateKey::LoadOrGenerate(const std::string& path) {
  if (fs::exists(path)) {
    return LoadFromFile(path);
  }
  auto key = Generate();
  key.SaveToFile(path);
  return key;
}

PublicKey PrivateKey::GetPublicKey() const {
  std::array<std::uint8_t, 32> pub{};
  crypto_scalarmult_base(pub.data(), bytes_.data());
  return PublicKey(pub);
}

std::string PrivateKey::ToBase64() const { return Base64Encode(bytes_); }

void PrivateKey::SaveToFile(const std::string& path) const {
  fs::path target(path);
  fs::path dir = target.parent_path();
  if (!dir.empty()) {
    fs::create_directories(dir);
#ifndef _WIN32
    ::chmod(dir.c_str(), 0700);
#endif
  }

  std::ofstream out(target, std::ios::trunc);
  if (!out) {
    throw Error(ErrorCode::kIo, "failed to write key file: " + path);
  }
  out << ToBase64() << "\n";
  out.close();
#ifndef _WIN32
  ::chmod(target.c_str(), 0600);
#endif
}

}  // namespace relayly
